internal/core/src/segcore/FieldIndexing.cpp

```diff
@@ -45,39 +49,140 @@ VectorFieldIndexing::BuildIndexRange(int64_t ack_beg,
     }
 }
 
+void
+VectorFieldIndexing::GetDataFromIndex(const int64_t* seg_offsets,
+                                int64_t count,
+                                int64_t element_size,
+                                void* output) {
+    auto ids_ds = std::make_shared<knowhere::DataSet>();
+    ids_ds->SetRows(count);
+    ids_ds->SetDim(1);
+    ids_ds->SetIds(seg_offsets);
+    ids_ds->SetIsOwner(false);
+
+    auto& vector = index_->GetVector(ids_ds);
+
+    std::memcpy(output, vector.data(), count * element_size);
+}
+
+void
+VectorFieldIndexing::AppendSegmentIndex(int64_t reserved_offset,
+                                        int64_t size,
+                                        const VectorBase* vec_base,
+                                        const void* data_source) {
+    AssertInfo(field_meta_.get_data_type() == DataType::VECTOR_FLOAT,
+               "Data type of vector field is not VECTOR_FLOAT");
+
+    auto dim = field_meta_.get_dim();
+    auto conf = get_build_params();
+    auto source = dynamic_cast<const ConcurrentVector<FloatVector>*>(vec_base);
+
+    auto per_chunk = source->get_size_per_chunk();
+    //append vector [vector_id_beg, vector_id_end] into index
+    //build index [vector_id_beg, build_threshold) when index not exist
+    if (!index_.get()) {
+        idx_t vector_id_beg = index_cur_.load();
+        idx_t vector_id_end = get_build_threshold() - 1;
+        auto chunk_id_beg =  vector_id_beg / per_chunk;
+        auto chunk_id_end =  vector_id_end / per_chunk;
+
+        int64_t vec_num = vector_id_end - vector_id_beg + 1;
+        // for train index
+        const void* data_addr;
+        std::unique_ptr<float[]> vec_data;
+        //all train data in one chunk
+        if (chunk_id_beg == chunk_id_end) {
+            data_addr = vec_base->get_chunk_data(chunk_id_beg);
+        } else {
+            //merge data from multiple chunks together
+            vec_data = std::make_unique<float[]>(vec_num * dim);
+            int64_t offset = 0;
+            //copy vector data [vector_id_beg, vector_id_end]
+            for (int chunk_id = chunk_id_beg; chunk_id <= chunk_id_end;
+                 chunk_id++) {
+                int chunk_offset = 0;
+                int chunk_copysz =
+                    chunk_id == chunk_id_end
+                        ? vector_id_end - chunk_id * per_chunk + 1
+                        : per_chunk;
+                std::memcpy(
+                    vec_data.get() + offset * dim,
+                    (const float*)vec_base->get_chunk_data(chunk_id) +
+                        chunk_offset * dim,
+                    chunk_copysz * dim * sizeof(float));
+                offset += chunk_copysz;
+            }
+            data_addr = vec_data.get();
+        }
+        auto dataset = knowhere::GenDataSet(vec_num, dim, data_addr);
```

Comment:
Will this method copy the data which data_addr points to?

Reply:
No copy here, dataset just shares the point.